File: analytics/ruby/example/lib/middleware/analytics.rb
require 'net/http'
require 'json'

module Middleware
  class Analytics
    def initialize(app)
      @app = app
    end

    def call(env)
      start = Time.now
      status, headers, response = @app.call(env)

      json = {
        api_key: "9e8820b1-6308-40f8-8423-d913c203b92d",
        hostname: env['HTTP_HOST'],
        path: env['REQUEST_PATH'],
        user_agent: env['HTTP_USER_AGENT'],
        method: env['REQUEST_METHOD'],
        status: status,
        framework: "Rails",
        response_time: (Time.now - start).to_f.round,
      }

      Thread.new {
        log_request(json)
      }

      [status, headers, response]
    end

    private

    def log_request(json)
      uri = URI('http://localhost:8080/api/log-request')
      http = Net::HTTP.new(uri.host, uri.port)
      Rails.logger.info("#{uri.host} #{uri.port} #{uri.path}")
      req = Net::HTTP::Post.new(uri.path, 'Content-Type' => 'application/json')
      req.body = json.to_json
      res = http.request(req)
    rescue => e
    end
  end
end

File: analytics/php/example/app/Http/Middleware/Analytics.php
<?php
 
namespace App\Http\Middleware;
 
use Closure;
 
class Analytics {

    public function __construct() {
        $this->api_key = getenv("ANALYTICS_API_KEY");
    }
    public function handle($request, Closure $next) {
        $start = microtime(true);
        $response = $next($request);

        $data = array(
            'api_key' => $this->api_key,
            'host' => $request->host(),
            'path' => $request->path(),
            'method' => $request->method(),
            'user_agent' => $request->userAgent(),
            'status' => $response->status(),
            'framework' => 'Laravel',
            'response_time' => (int) ($start - microtime(true)) * 1000,
        );

        $this->log_request($data);
        
        return $response;
    }
    
    private function log_request($data) {
        $url = "http://localhost:8080/api/log-request";
    
        $options = array(
                'http' => array(
                    'header'  => "Content-type: application/json\r\n",
                    'method'  => 'POST',
                    'content' => json_encode($data),
                    // 'timeout' => .01  // Don't wait for response
                )
        );
        
        try {
            $context = stream_context_create($options);
            if (($result = file_get_contents($url, false, $context)) === false) {
                var_dump($http_response_header);
            }
        }  catch (Exception $e) {
            // Fails silently
        }
    }
}